include/deal.II/lac/sparse_matrix.templates.h

```diff
@@ -1898,6 +1898,80 @@ SparseMatrix<number>::get_sparsity_pattern () const
 }
 
 
+namespace internal
+{
+  namespace SparseMatrix
+  {
+    template <class StreamType, typename realType>
+    inline
+    void print_value(StreamType &out, const realType&val)
+    {
+      out << val;
+    }
+
+    template<class StreamType, typename realType>
+    inline
+    void print_value(StreamType &out, const std::complex<realType> &val)
+    {
+      // the following format is what is required by the matrix market format.
+      out << val.real << ' ' << val.imag;
+    }
+
+    template<class StreamType, class numberType>
+    void
+    print_header(StreamType &out,const numberType & dummy)
+    {
+      out << "%%MatrixMarket matrix coordinate real general\n";
+    }
+
+    //template<typename realType>
+    template<class StreamType, class realType>
+    void
+    print_header(StreamType &out, const std::complex<realType> & dummy )
+    {
+      out << "%%MatrixMarket matrix coordinate complex general\n";
+    }  
+  } // namespace SparseMatrix
+} // namespace internal
+
+template <typename number>
+template <class StreamType>
+void SparseMatrix<number>::print_matrix_market(StreamType &out,
+					       const double threshold) const
+{
+  Assert(cols != 0, ExcNotInitialized());
+  Assert(val != 0, ExcNotInitialized());
+  Assert(threshold >= 0, ExcMessage("Negative threshold!") );  
+
+  //Print the header.
+  {
+    const number dummy = (number) 1.0; 
+    print_header(out,dummy);
```

Comment:
@davydden, In a lower reply @bangerth walked me through a better way to handle this.  Thanks.